### PR TITLE
Fix/mongo tests

### PR DESCRIPTION
**Pull request by ssalaues**
The mongo backend was found to be failing S3 functional in https://github.com/scality/S3/pull/1188

This fixes the failing versioning tests.

### CODE REVIEW

File: lib/storage/metadata/mongoclient/MongoClientInterface.js [hunk 1]
```diff
@@ -374,46 +374,65 @@ class MongoClientInterface {
     }
 
     /*
-     * In this case the aller provides a versionId. This function will
-     * sequentially update the object with given versionId THEN the
-     * master iff the provided versionId matches the one of the master
+     * In this case the caller provides a versionId. This function will
+     * check if the object exists then sequentially update the object
+     * (or create if doesn't exist) with given versionId THEN the master
+     * if the provided versionId matches the one of the master
```
Comment (vrancurel) [inline lib/storage/metadata/mongoclient/MongoClientInterface.js]: Please mention the race condition case

**Comment by ssalaues** [comment]
Add info on potential race condition

**Comment by ssalaues** [comment]
Changed the cb return for 11000 error to an internal error so that in the case of the race condition, retries can happen automatically and therefore prevent the race condition from impacting versionid put objects.

File: lib/storage/metadata/mongoclient/MongoClientInterface.js [hunk 2]
```diff
@@ -374,46 +374,72 @@ class MongoClientInterface {
...
+                        // case of the race condition mentioned above, the
+                        // InternalError will allow for automatic retries
+                        log.error('putObjectVerCase3:', errors.KeyAlreadyExist);
+                        return cb(errors.InternalError);
+                    }
+                    return cb(errors.NoSuchVersion);
+                }
+                return cb(null, `{"versionId": "${objVal.versionId}"}`);
+            });
+            return null;
```
Comment (dora-korpar) [inline lib/storage/metadata/mongoclient/MongoClientInterface.js]: Would be best practice to `return c.bulkWrite` instead of a null return

**Comment by ssalaues** [comment]
Changed the return to c.bulkWrite instead of the return null as recommended by @dora-korpar

File: lib/storage/metadata/mongoclient/MongoClientInterface.js [hunk 3]
```diff
@@ -374,46 +374,71 @@ class MongoClientInterface {
...
+            }, err => {
+                if (err) {
+                    log.error(
+                        'putObjectVerCase3: error putting object version',
+                    { error: err.message });
+                    if (err.code === 11000) {
+                        // We want duplicate key error logged however in
+                        // case of the race condition mentioned above, the
+                        // InternalError will allow for automatic retries
+                        log.error('putObjectVerCase3:', errors.KeyAlreadyExist);
```
Comment (rahulreddy) [inline lib/storage/metadata/mongoclient/MongoClientInterface.js]: There is a typo `KeyAlreadyExist` instead of KeyAlreadyExist**s**

**Pull request merged**